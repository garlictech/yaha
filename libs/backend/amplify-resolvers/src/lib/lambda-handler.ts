import { Context, Handler } from 'aws-lambda';
import {
  searchAroundHikeResolver,
  searchByRadiusResolver,
  searchInEnvelopeResolver,
  searchInMultipolygonResolver,
  searchInShapeResolver,
  searchSafeImagesAroundHikeResolver,
} from './search-resolvers';
const { createConnector } = require('aws-elasticsearch-js');
import { Client } from '@elastic/elasticsearch';
import { getGraphqlSdkForIAM, AmplifyApiConfig } from '@yaha/gql-api';

export interface AmplifyRequest {
  typeName: string;
  fieldName: string;
  identity?: {
    username?: string;
  };
  arguments: unknown;
}

const sdk = getGraphqlSdkForIAM(
  process.env.API_ACCESS_KEY_ID || '',
  process.env.API_SECRET_ACCESS_KEY || '',
);

const searchDeps = {
  osClient: new Client({
    nodes: [AmplifyApiConfig.openSearchEndpoint],
    Connection: createConnector({ region: process.env.AWS_REGION || '' }),
  }),
  sdk,
};

export const handler: Handler<AmplifyRequest, unknown> = (
  event: AmplifyRequest,
  _context: Context,
): Promise<unknown> => {
  console.debug('### Appsync Lambda handler ~ event:AmplifyRequest', event);

  // We don't use typing here, we explicitly check the fields in runtime.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const resolverMap: any = {
    Mutation: {},
    Query: {
      searchByRadius: searchByRadiusResolver(searchDeps),
      searchInShape: searchInShapeResolver(searchDeps),
      searchInEnvelope: searchInEnvelopeResolver(searchDeps),
      searchInMultipolygon: searchInMultipolygonResolver(searchDeps),
      searchAroundHike: searchAroundHikeResolver(searchDeps),
      searchSafeImagesAroundHike:
        searchSafeImagesAroundHikeResolver(searchDeps),
    },
  };

  const op = resolverMap[event.typeName]?.[event.fieldName]?.(event.arguments);

  if (op === undefined) {
    return Promise.reject(
      'Unknown graphql field in the appsync-lambda handler',
    );
  } else {
    return op.toPromise();
  }
};
