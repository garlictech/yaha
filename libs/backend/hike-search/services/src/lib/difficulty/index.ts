export { DifficultyFp } from './difficulty.fp';
