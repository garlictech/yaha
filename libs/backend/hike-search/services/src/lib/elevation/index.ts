export * from './elevation.service';
export * from './lib/elevation.fp';
