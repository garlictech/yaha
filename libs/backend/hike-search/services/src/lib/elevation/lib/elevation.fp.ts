import { Position } from '@turf/helpers';

const calculateHill = (
  data: number[][],
  bigEnough: (diff: number) => boolean,
  multiplier = 1,
): number => {
  let sum = 0;

  data.forEach((point, i) => {
    if (i > 0) {
      const diff = point[2] - data[i - 1][2];

      if (bigEnough(diff)) {
        sum += diff * multiplier;
      }
    }
  });

  return sum;
};

export class ElevationFp {
  static calculateUphill(data: Position[]): number {
    return calculateHill(data, diff => diff > 0);
  }

  static calculateDownhill(data: Position[]): number {
    return calculateHill(data, diff => diff < 0, -1);
  }
}
