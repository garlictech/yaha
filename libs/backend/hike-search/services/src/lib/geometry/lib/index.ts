export * from './circle-slice';
export * from './coordinate-iterator';
export * from './line-calculations';
export * from './point-transformations';
export * from './box-calculations';
export * from './polygon-calculations';
export * from './geospatial-calculations';
export * from './filter-points-close-to-reference-points';
export * from './hike-calculations';
export * from './circle-fp';
