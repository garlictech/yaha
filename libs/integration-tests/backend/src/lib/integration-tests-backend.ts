export function integrationTestsBackend(): string {
  return 'integration-tests-backend';
}
