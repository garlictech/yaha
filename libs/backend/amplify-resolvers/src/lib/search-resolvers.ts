import { YahaApi } from '@yaha/gql-api';
import { defer, from, map, Observable } from 'rxjs';
import { Client } from '@elastic/elasticsearch';
import * as R from 'ramda';
import { pipe } from 'fp-ts/lib/function';

export interface SearchResolverDeps {
  osClient: Client;
}

export const searchByRadiusResolver =
  (deps: SearchResolverDeps) =>
  (
    args: YahaApi.QuerySearchByRadiusArgs,
  ): Observable<YahaApi.GeoSearchConnection> =>
    pipe(
      {
        query: {
          bool: {
            must: {
              match_all: {},
            },
            filter: {
              geo_distance: {
                distance: `${args.query.radiusInMeters / 1000}km`,
                location: args.query.location,
              },
            },
          },
        },
        sort: [
          {
            _geo_distance: {
              location: args.query.location,
              order: 'asc',
              unit: 'km',
              distance_type: 'plane',
            },
            createdAt: 'desc',
          },
        ],
        size: args.query.limit ?? 10,
        track_total_hits: !args.query.nextToken,
      },
      body =>
        args.query.nextToken
          ? {
              ...body,
              search_after: JSON.parse(args.query.nextToken),
            }
          : body,
      body => ({
        index: args.query.objectType,
        body,
      }),
      x => defer(() => from(deps.osClient.search(x))),
      map(res =>
        pipe(
          res.body.hits?.hits ?? [],
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          R.map((hit: any) => hit?._source?.id as string),
          R.reject(x => R.isNil(x)),
          items => ({
            items,
            total: res.body.hits.total?.value,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            nextToken: (R.last(res.body.hits.hits) as any)?.sort
              ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
                JSON.stringify((R.last(res.body.hits.hits) as any).sort)
              : undefined,
          }),
        ),
      ),
    );

export const searchInShapeResolver =
  (deps: SearchResolverDeps) =>
  (
    args: YahaApi.QuerySearchInShapeArgs,
  ): Observable<YahaApi.GeoSearchConnection> =>
    pipe(
      {
        query: {
          bool: {
            must: {
              match_all: {},
            },
            filter: {
              geo_shape: {
                location: {
                  shape: {
                    ...args.query.shape,
                    coordinates: args.query.shape.coordinates,
                  },
                },
              },
            },
          },
        },
        size: args.query.limit ?? 10,
        track_total_hits: !args.query.nextToken,
      },
      R.tap(() => console.warn('*****', JSON.stringify(args, null, 2))),
      body =>
        args.query.nextToken
          ? {
              ...body,
              search_after: JSON.parse(args.query.nextToken),
            }
          : body,
      body => ({
        index: args.query.objectType,
        body,
      }),
      x => defer(() => from(deps.osClient.search(x))),
      map(res =>
        pipe(
          res.body.hits?.hits ?? [],
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          R.map((hit: any) => hit?._source?.id as string),
          R.reject(x => R.isNil(x)),
          items => ({
            items,
            total: res.body.hits.total?.value,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            nextToken: (R.last(res.body.hits.hits) as any)?.sort
              ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
                JSON.stringify((R.last(res.body.hits.hits) as any).sort)
              : undefined,
          }),
        ),
      ),
    );

export const searchInEnvelopeResolver =
  (deps: SearchResolverDeps) =>
  (
    args: YahaApi.QuerySearchInEnvelopeArgs,
  ): Observable<YahaApi.GeoSearchConnection> =>
    pipe(
      {
        query: {
          bool: {
            must: {
              match_all: {},
            },
            filter: {
              geo_shape: {
                location: {
                  shape: {
                    type: 'envelope',
                    coordinates: args.query.envelope,
                  },
                },
              },
            },
          },
        },
        size: args.query.limit ?? 10,
        track_total_hits: !args.query.nextToken,
      },
      body =>
        args.query.nextToken
          ? {
              ...body,
              search_after: JSON.parse(args.query.nextToken),
            }
          : body,
      body => ({
        index: args.query.objectType,
        body,
      }),
      x => defer(() => from(deps.osClient.search(x))),
      map(res =>
        pipe(
          res.body.hits?.hits ?? [],
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          R.map((hit: any) => hit?._source?.id as string),
          R.reject(x => R.isNil(x)),
          items => ({
            items,
            total: res.body.hits.total?.value,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            nextToken: (R.last(res.body.hits.hits) as any)?.sort
              ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
                JSON.stringify((R.last(res.body.hits.hits) as any).sort)
              : undefined,
          }),
        ),
      ),
    );

export const searchInMultipolygonResolver =
  (deps: SearchResolverDeps) =>
  (
    args: YahaApi.QuerySearchInMultipolygonArgs,
  ): Observable<YahaApi.GeoSearchConnection> =>
    pipe(
      {
        query: {
          bool: {
            must: {
              match_all: {},
            },
            filter: {
              geo_shape: {
                location: {
                  shape: {
                    type: 'multipolygon',
                    coordinates: args.query.coordinates,
                  },
                },
              },
            },
          },
        },
        size: args.query.limit ?? 10,
        track_total_hits: !args.query.nextToken,
      },
      body =>
        args.query.nextToken
          ? {
              ...body,
              search_after: JSON.parse(args.query.nextToken),
            }
          : body,
      body => ({
        index: args.query.objectType,
        body,
      }),
      x => defer(() => from(deps.osClient.search(x))),
      map(res =>
        pipe(
          res.body.hits?.hits ?? [],
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          R.map((hit: any) => hit?._source?.id as string),
          R.reject(x => R.isNil(x)),
          items => ({
            items,
            total: res.body.hits.total?.value,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            nextToken: (R.last(res.body.hits.hits) as any)?.sort
              ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
                JSON.stringify((R.last(res.body.hits.hits) as any).sort)
              : undefined,
          }),
        ),
      ),
    );
