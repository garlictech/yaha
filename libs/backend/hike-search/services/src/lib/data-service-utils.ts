import { from, Observable } from 'rxjs';
import { concatMap, mergeMap, toArray } from 'rxjs/operators';
import { pipe } from 'fp-ts/lib/function';
import { throwIfEmptyValue } from '@yaha/shared/utils';

export const multipleGet =
  <ITEM>(getOp: ({ id }: { id: string }) => Observable<ITEM>) =>
  (ids: string[]): Observable<ITEM[]> =>
    pipe(
      from(ids),
      mergeMap(id => getOp({ id }), 5),
      throwIfEmptyValue(),
      toArray(),
    );

export const multipleWrite =
  <CREATE, ITEM>(
    createOp: ({
      input,
    }: {
      input: CREATE;
    }) => Observable<ITEM | undefined | null>,
  ) =>
  (items: CREATE[]): Observable<Array<ITEM | undefined | null>> =>
    pipe(
      from(items),
      concatMap(item => createOp({ input: item })),
      throwIfEmptyValue(),
      toArray(),
    );
