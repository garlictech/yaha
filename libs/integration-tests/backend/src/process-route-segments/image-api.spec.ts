import { NestFactory } from '@nestjs/core';
import * as E from 'fp-ts/lib/Either';
import { tap, map, switchMap } from 'rxjs/operators';
import {
  ImageApiService,
  ImageModule,
} from '@bit/garlictech.nestjs.gtrack.image';
import { Image } from '@bit/garlictech.universal.gtrack.graphql-api';

import { imageFixture, getIdOfImage } from './fixtures';

describe('Image graphql integration tests', () => {
  let imageApiService: ImageApiService;

  beforeAll(async () => {
    const app = await NestFactory.createApplicationContext(ImageModule);
    imageApiService = app.get(ImageApiService);
  });

  test('it must be able to CRUD images', done => {
    imageApiService.api
      .delete(getIdOfImage(imageFixture()))
      .pipe(
        switchMap(() =>
          imageApiService.api.createMultipleItems([imageFixture()]),
        ),
        tap(result => expect(result).toMatchSnapshot()),
        map(E.fold(done.fail, (res: Image[]) => res[0].id)),
        switchMap((id: string) => imageApiService.api.getById(id)),
        tap(result => expect(result).toMatchSnapshot()),
        map(E.fold(done.fail, (res: Image) => res.id)),
        switchMap((id: string) =>
          imageApiService.api.getMultipleItemsById([id]),
        ),
        tap(result => expect(result).toMatchSnapshot()),
        map(E.fold(done.fail, res => res[0].id)),
        switchMap((id: string) =>
          imageApiService.api.update(id, {
            attributions: 'foobar attribution',
          }),
        ),
        tap(result => expect(result).toMatchSnapshot()),
        map(E.fold(done.fail, (res: Image) => res.id)),
        switchMap((id: string) => imageApiService.api.delete(id)),
        tap(result => expect(result).toMatchSnapshot()),
      )
      .subscribe(
        () => done(),
        err => done.fail(err),
      );
  }, 20000);
});
