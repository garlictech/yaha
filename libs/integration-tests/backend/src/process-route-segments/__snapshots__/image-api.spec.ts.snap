// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image graphql integration tests it must be able to CRUD images 1`] = `
Object {
  "_tag": "Right",
  "right": Array [
    Object {
      "__typename": "Image",
      "attributions": null,
      "card": Object {
        "__typename": "ImageInfo",
        "height": 12,
        "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_z.jpg",
        "width": 640,
      },
      "id": "flickr-8145547906",
      "labels": null,
      "lat": 47.481277,
      "lon": 18.907341,
      "original": Object {
        "__typename": "ImageInfo",
        "height": 12,
        "url": "https://live.staticflickr.com/8471/8145547906_bf5900cb6f_o.jpg",
        "width": 2592,
      },
      "sourceObject": Object {
        "__typename": "PoiSourceObject",
        "objectId": "8145547906",
        "objectType": "flickr",
      },
      "thumbnail": Object {
        "__typename": "ImageInfo",
        "height": 12,
        "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_n.jpg",
        "width": 320,
      },
    },
  ],
}
`;

exports[`Image graphql integration tests it must be able to CRUD images 2`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "__typename": "Image",
    "attributions": null,
    "card": Object {
      "__typename": "ImageInfo",
      "height": 12,
      "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_z.jpg",
      "width": 640,
    },
    "id": "flickr-8145547906",
    "labels": null,
    "lat": 47.481277,
    "lon": 18.907341,
    "original": Object {
      "__typename": "ImageInfo",
      "height": 12,
      "url": "https://live.staticflickr.com/8471/8145547906_bf5900cb6f_o.jpg",
      "width": 2592,
    },
    "sourceObject": Object {
      "__typename": "PoiSourceObject",
      "objectId": "8145547906",
      "objectType": "flickr",
    },
    "thumbnail": Object {
      "__typename": "ImageInfo",
      "height": 12,
      "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_n.jpg",
      "width": 320,
    },
  },
}
`;

exports[`Image graphql integration tests it must be able to CRUD images 3`] = `
Object {
  "_tag": "Right",
  "right": Array [
    Object {
      "__typename": "Image",
      "attributions": null,
      "card": Object {
        "__typename": "ImageInfo",
        "height": 12,
        "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_z.jpg",
        "width": 640,
      },
      "id": "flickr-8145547906",
      "labels": null,
      "lat": 47.481277,
      "lon": 18.907341,
      "original": Object {
        "__typename": "ImageInfo",
        "height": 12,
        "url": "https://live.staticflickr.com/8471/8145547906_bf5900cb6f_o.jpg",
        "width": 2592,
      },
      "sourceObject": Object {
        "__typename": "PoiSourceObject",
        "objectId": "8145547906",
        "objectType": "flickr",
      },
      "thumbnail": Object {
        "__typename": "ImageInfo",
        "height": 12,
        "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_n.jpg",
        "width": 320,
      },
    },
  ],
}
`;

exports[`Image graphql integration tests it must be able to CRUD images 4`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "__typename": "Image",
    "attributions": "foobar attribution",
    "card": Object {
      "__typename": "ImageInfo",
      "height": 12,
      "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_z.jpg",
      "width": 640,
    },
    "id": "flickr-8145547906",
    "labels": null,
    "lat": 47.481277,
    "lon": 18.907341,
    "original": Object {
      "__typename": "ImageInfo",
      "height": 12,
      "url": "https://live.staticflickr.com/8471/8145547906_bf5900cb6f_o.jpg",
      "width": 2592,
    },
    "sourceObject": Object {
      "__typename": "PoiSourceObject",
      "objectId": "8145547906",
      "objectType": "flickr",
    },
    "thumbnail": Object {
      "__typename": "ImageInfo",
      "height": 12,
      "url": "https://live.staticflickr.com/8471/8145547906_0786c3d50c_n.jpg",
      "width": 320,
    },
  },
}
`;

exports[`Image graphql integration tests it must be able to CRUD images 5`] = `
Object {
  "_tag": "Right",
  "right": true,
}
`;
